mithril-client/src/utils/ancillary_verifier.rs

```diff
@@ -0,0 +1,395 @@
+}
+
+#[cfg(test)]
+mod tests {
+    use std::io::Write;
+
+    use mithril_common::temp_dir_create;
+
+    use super::*;
+
+    fn write_file(target_path: &Path, content: &str) {
+        let mut file = File::create(target_path).unwrap();
+        write!(file, "{content}").unwrap();
+    }
+
+    mod ancillary_verifier {
+        use std::collections::BTreeMap;
+
+        use mithril_common::{crypto_helper::ManifestSigner, test_utils::fake_keys};
+        use sha2::{Digest, Sha256};
+
+        use super::*;
+
+        fn write_manifest(target_dir: &Path, manifest: &AncillaryFilesManifest) {
+            serde_json::to_writer(
+                File::create(target_dir.join(AncillaryFilesManifest::ANCILLARY_MANIFEST_FILE_NAME))
+                    .unwrap(),
+                manifest,
+            )
+            .unwrap();
+        }
+
+        fn generate_manifest_item(
+            target_folder: &Path,
+            filename: &str,
+            content: &str,
+        ) -> (PathBuf, String) {
+            write_file(&target_folder.join(filename), content);
+            (
+                PathBuf::from(filename),
+                compute_sha256_hash(content.as_bytes()),
+            )
+        }
+
+        fn compute_sha256_hash(data: impl AsRef<[u8]>) -> String {
+            hex::encode(Sha256::digest(data))
+        }
+
+        #[tokio::test]
+        async fn fail_if_no_manifest_found() {
+            let temp_dir = temp_dir_create!();
+
+            let ancillary_verifier = AncillaryVerifier::new(
+                fake_keys::manifest_verification_key()[0]
+                    .try_into()
+                    .unwrap(),
+            );
+
+            let err = ancillary_verifier.verify(&temp_dir).await.unwrap_err();
+            assert!(
+                matches!(err, AncillaryVerificationError::ManifestParse(..)),
+                "Expected AncillaryVerificationError::ManifestParse, got: {err:?}",
+            );
+        }
+
+        #[tokio::test]
+        async fn fail_if_manifest_is_not_a_valid_json() {
+            let temp_dir = temp_dir_create!();
+            let manifest_file =
+                File::create(temp_dir.join(AncillaryFilesManifest::ANCILLARY_MANIFEST_FILE_NAME))
+                    .unwrap();
+            write!(&manifest_file, "invalid json").unwrap();
+
+            let ancillary_verifier = AncillaryVerifier::new(
+                fake_keys::manifest_verification_key()[0]
+                    .try_into()
+                    .unwrap(),
+            );
+
+            let err = ancillary_verifier.verify(&temp_dir).await.unwrap_err();
+            assert!(
+                matches!(err, AncillaryVerificationError::ManifestParse(..)),
+                "Expected AncillaryVerificationError::ManifestParse, got: {err:?}",
+            );
+        }
+
+        #[tokio::test]
+        async fn fail_if_manifest_is_invalid() {
+            let temp_dir = temp_dir_create!();
+            let manifest = AncillaryFilesManifest {
```

Comment:
We can create a function to create a valid manifest and only change one thing to make it invalid.
```
let manifest = AncillaryFilesManifest {
                data: BTreeMap::from([(PathBuf::from("invalid path"), "invalid sha".to_string())]),
                ..valid_manifest()
            };
```